litgpt/data/text_files.py

```diff
@@ -0,0 +1,161 @@
+import glob
+import os
+from dataclasses import dataclass, field
+from functools import partial
+from pathlib import Path
+from tqdm import tqdm
+from typing import Optional
+
+from torch.utils.data import DataLoader
+
+from litgpt import Tokenizer
+from litgpt.data import DataModule
+
+
+def get_half_workers():
+    num_workers = (os.cpu_count() - 1) // 2
+    if num_workers < 1:
+        return 1
+    else:
+        return num_workers
+
+
+@dataclass
+class TextFiles(DataModule):
+    """The TextFile data module used for pretraining.
+
+    Reads in text data from plaintext files contained in a data folder
+    and provides training and validation dataloaders that return batches of tokens.
+    Every sample is set to a fixed length.
+    """
+    train_data_path: Path
+    """The path to the data directory used for training that contains .txt files"""
+    val_data_path: Optional[str] = None
+    """The path to the data directory used for validation that
+    contains .txt files. Splits off data for validation from the
+    training set if None."""
+    seed: int = 42
+    """The seed to use for shuffling the dataset."""
+    num_workers: Optional[int] = None
+    """The number of workers to use for data loading.
+       Sets the number of workers equal to the number of avaialable CPUs-1 by default."""
+
+    tokenizer: Optional[Tokenizer] = field(default=None, init=False, repr=False)
+    batch_size: int = field(default=1, init=False, repr=False)
+    max_seq_length: int = field(default=-1, init=False, repr=False)
+
+    def connect(self, tokenizer: Optional[Tokenizer] = None, batch_size: int = 1, max_seq_length: int = -1) -> None:
+        self.tokenizer = tokenizer
+        self.batch_size = batch_size
+        self.max_seq_length = max_seq_length + 1  # Increase by one because we need the next token as well
+
+    def __post_init__(self) -> None:
+        self.out_path_train = self.train_data_path / "train"
+        if self.val_data_path is None:
+            self.out_path_val = self.train_data_path / "val"
+        else:
+            self.out_path_val = Path(self.val_data_path) / "val"
+
+    def prepare_data(self) -> None:
+        from litdata import optimize
+
+        train_files = sorted(glob.glob(str(self.train_data_path / "*.txt")))
+        assert len(train_files) > 0, f"No .txt files found in train data {train_files}"
+        assert len(train_files) > 1, f"Expected at least two .txt files in {train_files}"
+
+        if self.val_data_path is not None:
+            self.val_data_path = Path(self.val_data_path)
+            val_files = sorted(glob.glob(str(self.val_data_path / "*.txt")))
+            assert len(val_files) > 0, f"No .txt files found in validation data {val_files}"
+        # train/test split. let's use only shard 0 for test split, rest train
+        else:
+            val_files, *train_files = train_files
+            val_files = [val_files] if not isinstance(val_files, list) else val_files
```

Comment:
How could val_files be a list already?

```suggestion
            val_files = [val_files]
```